CT: Load log server configuration from CS.cfg

SilleBille opened the pull request:
This patch:
- Loads CT log server configurations from CS.cfg
- Can submit precertificates to multiple log servers
- Creates CT relevant java packages

### TODO:
Refactor code to:
- Embed multiple SCT responses onto certificate
- accommodate REST structure
- Move CT related code into it's related java package

`Signed-off-by: Dinesh Prasanth M K <dmoluguw@redhat.com>`

## Demo

#### CS.cfg
````
ca.certTransparency.enable=true
ca.certTransparency.log.1.enable=true
ca.certTransparency.log.1.pubKey=MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEw8i8S7qiGEs9NXv0ZJFh6uuOmR2Q7dPprzk9XNNGkUXjzqx2SDvRfiwKYwBljfWujozHESVPQyydGaHhkaSz/g==
ca.certTransparency.log.1.url=http://ct.googleapis.com:80/testtube/
ca.certTransparency.log.1.version=1
ca.certTransparency.log.2.enable=true
ca.certTransparency.log.2.pubKey=MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEKATl2B3SAbxyzGOfNRB+AytNTGvdF/FFY6HzWb+/HPE4lJ37vx2nEm99KYUy9SoNzF5VyTwCQG5nL/c5Q77yQQ==
ca.certTransparency.log.2.url=http://ct.googleapis.com:80/logs/crucible/
ca.certTransparency.log.2.version=1
ca.certTransparency.log.3.enable=false
ca.certTransparency.log.3.pubKey=MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEiKfWtuoWCPMEzSKySjMjXpo38WOdZr6Yq0WYa2JQOv1uVMxkqHywf9Gz1kGeRLq/Rz3tVVvXgqb4jQ1UqKVKnw==
ca.certTransparency.log.3.url=http://ct.googleapis.com:80/logs/solera2020/
ca.certTransparency.log.3.version=1
ca.certTransparency.log.num=3
````

#### Debug log
````
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] INFO: CA is going to sign certificate serial number 0x15
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService: issueX509Cert: setting issuerDN using exact CA signing cert subjectDN encoding
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService: issueX509Cert - CT: about to check CT poison
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService: issueX509Cert - CT: ctPoison found
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService: issueX509Cert - CT: About to ca.sign CT pre-cert.
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: sign cert get algorithm
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: sign cert encoding cert
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: sign cert encoding algorithm
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CA cert signing: signing cert
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Getting algorithm context for SHA256withRSA RSASignatureWithSHA256Digest
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Signing Certificate
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Property ca.signing.testSignatureFailure not found
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.signing.testSignatureFailure=false
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService.createCTRequestcaCertInChain 0 = MIIDyjCCArKgAwIBAgIBDjANBgkqhkiG9w0BAQsFADBYMRcwFQYDVQQKDA5ESU5FU0hQUkFTQU5USDETMBEGA1UECwwKcGtpLXRvbWNhdDEoMCYGA1UEAwwfREVWIFRFU1QgQ0EgU2lnbmluZyBDZXJ0aWZpY2F0ZTAeFw0yMDA0MzAxNTM0MTZaFw0zOTA1MjkxOTM1MjVaMEgxEDAOBgNVBAoTB0VYQU1QTEUxEzARBgNVBAsTCnBraS10b21jYXQxHzAdBgNVBAMTFkNBIFNpZ25pbmcgQ2VydGlmaWNhdGUwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDGL6EmbVWME/5z5CigigFK1sFklHYrQ77Nu1QgAuGL+o1n8TU08qRasV9Dw92Yj00yYXXLoAjnyHkcq8g6Nhe3P50oFNiM/4zG6OMqZaziUgIFri6qEtf0xZNr5G2GCgK3fg+9Em8dqSx85iB5Nm93xsRO3xYyVju5G/IkO2sVXsKlKydHw3BG6DqFNIh9nsQH++RTXhA5WrybbrYm0s1rh/ADLiR/ZnhT/XwzUdO4JsAhH9OFCDK4Mhhh6cGXiXh0Wh1TuHID8cgxGfgtSHwhfYev3kuYylFr7mhiUCxAplZYxXXBCXpsbS86Jd3zaKfK4T28uboRSozhBByDXlmfAgMBAAGjga4wgaswHwYDVR0jBBgwFoAUgGDAgdpM2MyH/00/kXQzHufgXywwHQYDVR0OBBYEFMiodNd/9iRt6RsKmn77T5efzsMBMA8GA1UdEwEB/wQFMAMBAf8wDgYDVR0PAQH/BAQDAgHGMEgGCCsGAQUFBwEBBDwwOjA4BggrBgEFBQcwAYYsaHR0cDovL2N0dGVzdC5kaW5lc2hwcmFzYW50aC5tZTo4MDgwL2NhL29jc3AwDQYJKoZIhvcNAQELBQADggEBAG8c1di5FIM705wKgnwu/mDiHJ1hpaQaT6oTSLWfFGHkvkb/k2kgMXuSR9IsnD5U0+FTDbqUMaY1s3oZXnZFRIBnywPSpisaShBLZardIAoe6kX3N/tAhv3IE/T22ikoaRkfpcJB+yMdKHbR8/B24eDoa89Si60cOrYRWDRI/rYby3UHtM4WXpXJvY2GJX+UFnPByOOXKbK2a8RcjMX62rOr4tkaEl1oiZWUpRPIi8t96pIiETO0mWpQEUtEQk5qp3doyu9kVGEpzVj+JeSkVXjFY66pvgpfIPu7+tF8e6PY6EnAhYUDM/3uSiU33H9TnrWy2lGeLvk2H1GVQg6byBY=
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService.createCTRequestcaCertInChain 1 = MIID2jCCAsKgAwIBAgIBATANBgkqhkiG9w0BAQsFADBYMRcwFQYDVQQKDA5ESU5FU0hQUkFTQU5USDETMBEGA1UECwwKcGtpLXRvbWNhdDEoMCYGA1UEAwwfREVWIFRFU1QgQ0EgU2lnbmluZyBDZXJ0aWZpY2F0ZTAeFw0xOTA1MjkxOTM1MjVaFw0zOTA1MjkxOTM1MjVaMFgxFzAVBgNVBAoMDkRJTkVTSFBSQVNBTlRIMRMwEQYDVQQLDApwa2ktdG9tY2F0MSgwJgYDVQQDDB9ERVYgVEVTVCBDQSBTaWduaW5nIENlcnRpZmljYXRlMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyKTONtA0xxcRCD1UbGe8NBOwkaCpjucrqZXBmOzgqrdRN+ovR5mZVtaZN4lF+Z1Qnu/VcK/VX2BSsVJDwboa74NuEsyY1cs8Yimu3T0KbVRXvNVGmFMYCsrSQ4ap8/loiayiybH0mcj12zmXm8UDM41PsX/3FrXqHAaGRQCSMYRTIsM1GiKnM5dvY15geAKpxPJrUwWxgu1YbkHwATWM6IZqhexc+MWHOQ5C4CidmObuB6Oku/uoWNa+y3Ea/htzzIoNsy+Up5O3+fUgV12LcMkdj7yS7RpmmR8xlxeI6slcE/kTYkAw1f2qvxH4xv3MlroKy7PZ+qoHDtHuWuiToQIDAQABo4GuMIGrMB8GA1UdIwQYMBaAFIBgwIHaTNjMh/9NP5F0Mx7n4F8sMA8GA1UdEwEB/wQFMAMBAf8wDgYDVR0PAQH/BAQDAgHGMB0GA1UdDgQWBBSAYMCB2kzYzIf/TT+RdDMe5+BfLDBIBggrBgEFBQcBAQQ8MDowOAYIKwYBBQUHMAGGLGh0dHA6Ly9jdHRlc3QuZGluZXNocHJhc2FudGgubWU6ODA4MC9jYS9vY3NwMA0GCSqGSIb3DQEBCwUAA4IBAQCGzHPk61ew00PC5TaqEhB2+fXbHb2RAPHzFaFQSREkX8B7MxCvJo9ylc5Tk7GYL45GjdJLwbBp339Yz116muNCf4VJidvRa52BiGvUypKylfclLAB6Uk2h53IR2Sgioy2d63WcvQEl0VoK32dVYLsvTEmKl0CRa+oaCUHOOUa/zM95NHOWFbak92NBRpdd/lsZ8lqy1AUENOjnpvyxUHsfq8p1Rwao9ED7Zj7LQr7L7p6ZLLbjzsb59/5kin876wTE8NV3ZWZRhUy6IjSG7Mvj9cGWKzvIGerl4X1DgykBf03nwXn4Nq+pQavYcMvrejrM41Vy90G/0MfX8+xtaAXQ
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService.createCTRequest ct_json_request:{"chain":["MIIDsTCCApmgAwIBAgIBFTANBgkqhkiG9w0BAQsFADBIMRAwDgYDVQQKEwdFWEFNUExFMRMwEQYDVQQLEwpwa2ktdG9tY2F0MR8wHQYDVQQDExZDQSBTaWduaW5nIENlcnRpZmljYXRlMB4XDTIwMDUwNzE3MzkzMloXDTIyMDQyNzE3MzkzMlowIDEeMBwGA1UEAwwVc2lsbGViaWxsZS50ZXN0Y3QuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxEgSNw1ZEaCHxF5FOmjBQBobHhFaDkwa0qCveq2isLP6EMVnzfyr3hxfMFNbsAI2p8X/OtAM9HJr1eZcSNIIfPRcqkhA9OtKpztnzdqm6+1eGspYRhKzww+4T5TspdO4fivn9AvszHRiN8YfQewbtdEUF+fzqFmg9THa1VrN2Xdh0ci868tb4E++ZM27yejNkpGTp9n9nV7/xQm8O2DOOTIeXHfz+s1flGdYa8fi+KSuhuTyNXbm4+6v9P4tYQ8CXI2jQDtdda5kTEu0Q4ws01GmEULdMiDgQa2/3akummZZu8z7FWC3CKkfgbde5r5Vx2dxun2MzqAokbKZ38XZIQIDAQABo4HNMIHKMB8GA1UdIwQYMBaAFMiodNd/9iRt6RsKmn77T5efzsMBMBMGCisGAQQB1nkCBAMBAf8EAgUAMCAGA1UdEQQZMBeCFXNpbGxlYmlsbGUudGVzdGN0LmNvbTBBBggrBgEFBQcBAQQ1MDMwMQYIKwYBBQUHMAGGJWh0dHA6Ly9ta2QuZmVkb3JhLm9mZmljZTo4MDgwL2NhL29jc3AwDgYDVR0PAQH/BAQDAgSwMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjANBgkqhkiG9w0BAQsFAAOCAQEAPfXln4Mb9qkA3IG6A3sp7jBi8VvMlP8E87uUYHsLMGs0FRGbHtGb0qjEKmQ2LRM4FYSWSvIkcg5vnR0++nLz1f9enftob/Na9Yi9Ap4c8c7x15sbuHbTLolgZGx4rQF9V2YiVuvsTvgByhS7J9Ve7SmbqntO57q13PcDbPmW8rIdQoGrEn9AUfbuvmkB29wkPSspZ7Yx1pC4d+X4fV5OSzXEGeOZevA142M9pVa7Es7eBPA3pqlWQhGXafBGGlxKgFRYIzzq1q8vB1sTm267I6uAHoa0/T1gQlPEz9Q5CphwXknl9/wySRdj6+lUI/7/SYjO8K7C0mJBjXTojy9+Mg==","MIIDyjCCArKgAwIBAgIBDjANBgkqhkiG9w0BAQsFADBYMRcwFQYDVQQKDA5ESU5FU0hQUkFTQU5USDETMBEGA1UECwwKcGtpLXRvbWNhdDEoMCYGA1UEAwwfREVWIFRFU1QgQ0EgU2lnbmluZyBDZXJ0aWZpY2F0ZTAeFw0yMDA0MzAxNTM0MTZaFw0zOTA1MjkxOTM1MjVaMEgxEDAOBgNVBAoTB0VYQU1QTEUxEzARBgNVBAsTCnBraS10b21jYXQxHzAdBgNVBAMTFkNBIFNpZ25pbmcgQ2VydGlmaWNhdGUwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDGL6EmbVWME/5z5CigigFK1sFklHYrQ77Nu1QgAuGL+o1n8TU08qRasV9Dw92Yj00yYXXLoAjnyHkcq8g6Nhe3P50oFNiM/4zG6OMqZaziUgIFri6qEtf0xZNr5G2GCgK3fg+9Em8dqSx85iB5Nm93xsRO3xYyVju5G/IkO2sVXsKlKydHw3BG6DqFNIh9nsQH++RTXhA5WrybbrYm0s1rh/ADLiR/ZnhT/XwzUdO4JsAhH9OFCDK4Mhhh6cGXiXh0Wh1TuHID8cgxGfgtSHwhfYev3kuYylFr7mhiUCxAplZYxXXBCXpsbS86Jd3zaKfK4T28uboRSozhBByDXlmfAgMBAAGjga4wgaswHwYDVR0jBBgwFoAUgGDAgdpM2MyH/00/kXQzHufgXywwHQYDVR0OBBYEFMiodNd/9iRt6RsKmn77T5efzsMBMA8GA1UdEwEB/wQFMAMBAf8wDgYDVR0PAQH/BAQDAgHGMEgGCCsGAQUFBwEBBDwwOjA4BggrBgEFBQcwAYYsaHR0cDovL2N0dGVzdC5kaW5lc2hwcmFzYW50aC5tZTo4MDgwL2NhL29jc3AwDQYJKoZIhvcNAQELBQADggEBAG8c1di5FIM705wKgnwu/mDiHJ1hpaQaT6oTSLWfFGHkvkb/k2kgMXuSR9IsnD5U0+FTDbqUMaY1s3oZXnZFRIBnywPSpisaShBLZardIAoe6kX3N/tAhv3IE/T22ikoaRkfpcJB+yMdKHbR8/B24eDoa89Si60cOrYRWDRI/rYby3UHtM4WXpXJvY2GJX+UFnPByOOXKbK2a8RcjMX62rOr4tkaEl1oiZWUpRPIi8t96pIiETO0mWpQEUtEQk5qp3doyu9kVGEpzVj+JeSkVXjFY66pvgpfIPu7+tF8e6PY6EnAhYUDM/3uSiU33H9TnrWy2lGeLvk2H1GVQg6byBY=","MIID2jCCAsKgAwIBAgIBATANBgkqhkiG9w0BAQsFADBYMRcwFQYDVQQKDA5ESU5FU0hQUkFTQU5USDETMBEGA1UECwwKcGtpLXRvbWNhdDEoMCYGA1UEAwwfREVWIFRFU1QgQ0EgU2lnbmluZyBDZXJ0aWZpY2F0ZTAeFw0xOTA1MjkxOTM1MjVaFw0zOTA1MjkxOTM1MjVaMFgxFzAVBgNVBAoMDkRJTkVTSFBSQVNBTlRIMRMwEQYDVQQLDApwa2ktdG9tY2F0MSgwJgYDVQQDDB9ERVYgVEVTVCBDQSBTaWduaW5nIENlcnRpZmljYXRlMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyKTONtA0xxcRCD1UbGe8NBOwkaCpjucrqZXBmOzgqrdRN+ovR5mZVtaZN4lF+Z1Qnu/VcK/VX2BSsVJDwboa74NuEsyY1cs8Yimu3T0KbVRXvNVGmFMYCsrSQ4ap8/loiayiybH0mcj12zmXm8UDM41PsX/3FrXqHAaGRQCSMYRTIsM1GiKnM5dvY15geAKpxPJrUwWxgu1YbkHwATWM6IZqhexc+MWHOQ5C4CidmObuB6Oku/uoWNa+y3Ea/htzzIoNsy+Up5O3+fUgV12LcMkdj7yS7RpmmR8xlxeI6slcE/kTYkAw1f2qvxH4xv3MlroKy7PZ+qoHDtHuWuiToQIDAQABo4GuMIGrMB8GA1UdIwQYMBaAFIBgwIHaTNjMh/9NP5F0Mx7n4F8sMA8GA1UdEwEB/wQFMAMBAf8wDgYDVR0PAQH/BAQDAgHGMB0GA1UdDgQWBBSAYMCB2kzYzIf/TT+RdDMe5+BfLDBIBggrBgEFBQcBAQQ8MDowOAYIKwYBBQUHMAGGLGh0dHA6Ly9jdHRlc3QuZGluZXNocHJhc2FudGgubWU6ODA4MC9jYS9vY3NwMA0GCSqGSIb3DQEBCwUAA4IBAQCGzHPk61ew00PC5TaqEhB2+fXbHb2RAPHzFaFQSREkX8B7MxCvJo9ylc5Tk7GYL45GjdJLwbBp339Yz116muNCf4VJidvRa52BiGvUypKylfclLAB6Uk2h53IR2Sgioy2d63WcvQEl0VoK32dVYLsvTEmKl0CRa+oaCUHOOUa/zM95NHOWFbak92NBRpdd/lsZ8lqy1AUENOjnpvyxUHsfq8p1Rwao9ED7Zj7LQr7L7p6ZLLbjzsb59/5kin876wTE8NV3ZWZRhUy6IjSG7Mvj9cGWKzvIGerl4X1DgykBf03nwXn4Nq+pQavYcMvrejrM41Vy90G/0MfX8+xtaAXQ"]}
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.certTransparency.log.num=3
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Loading configuration for logserver ID: 1
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: logserver enabled: true
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.certTransparency.log.1.pubKey=MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEw8i8S7qiGEs9NXv0ZJFh6uuOmR2Q7dPprzk9XNNGkUXjzqx2SDvRfiwKYwBljfWujozHESVPQyydGaHhkaSz/g==
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.certTransparency.log.1.version=1
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.certTransparency.log.1.url=http://ct.googleapis.com:80/testtube/
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Loading configuration for logserver ID: 2
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: logserver enabled: true
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.certTransparency.log.2.pubKey=MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEKATl2B3SAbxyzGOfNRB+AytNTGvdF/FFY6HzWb+/HPE4lJ37vx2nEm99KYUy9SoNzF5VyTwCQG5nL/c5Q77yQQ==
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.certTransparency.log.2.version=1
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.certTransparency.log.2.url=http://ct.googleapis.com:80/logs/crucible/
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Loading configuration for logserver ID: 3
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: logserver enabled: false
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] INFO: Logserver ID: 3 has been disabled. Skipping this specific logserver
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Processing log server ID: 1
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Log server host: ct.googleapis.com
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Log server port: 80
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: Log server URI: http://ct.googleapis.com:80/testtube/ct/v1/add-pre-chain
2020-05-07 13:39:47 [https-jsse-nio-8443-exec-15] FINE: CAService.certTransSendReq begins
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: version HTTP/1.0
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: status code 200
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: reason OK
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: content {"sct_version":0,"id":"sMyD5aX5fWuvfAnMKEkEhyrH6IsTLGNQt8b9JuFsbHc=","timestamp":1588873188085,"extensions":"","signature":"BAMARzBFAiBTR/aOcihnDNKVXYLNtdWJejYTDJlTYn5/oN0BEGnRxQIhAPMXuGAqdKWIwJW6Gtup1gF5iBQs4uIwv/yySTKNmw32"}
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: CAService.certTransSendReq ends
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: Response from log server {"sct_version":0,"id":"sMyD5aX5fWuvfAnMKEkEhyrH6IsTLGNQt8b9JuFsbHc=","timestamp":1588873188085,"extensions":"","signature":"BAMARzBFAiBTR/aOcihnDNKVXYLNtdWJejYTDJlTYn5/oN0BEGnRxQIhAPMXuGAqdKWIwJW6Gtup1gF5iBQs4uIwv/yySTKNmw32"}
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT: begins
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT:  ct_timestamp: 00000171.f038d2f5
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: timeStampHexStringToByteArray: len =17
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT:  issuer_key_hash.length = 32
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT:  data_len = 715
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT: failed to verify SCT signature
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: verifySCT ends
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: Processing log server ID: 2
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: Log server host: ct.googleapis.com
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: Log server port: 80
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: Log server URI: http://ct.googleapis.com:80/logs/crucible/ct/v1/add-pre-chain
2020-05-07 13:39:48 [https-jsse-nio-8443-exec-15] FINE: CAService.certTransSendReq begins
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: version HTTP/1.0
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: status code 200
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: reason OK
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: content {"sct_version":0,"id":"w78Dp+HKiEHGB7rj/0Jw/KXsRbGG675OLPP8d4Yw9fY=","timestamp":1588873188673,"extensions":"","signature":"BAMARzBFAiANLSxiimhACoGuUttVPn0V/pP8O7jtYpZowZSmivty1AIhAPeZXrGMOJBrlGjmH41qab7qAnwFBdhVg8Dx7W6Qfv0E"}
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService.certTransSendReq ends
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: Response from log server {"sct_version":0,"id":"w78Dp+HKiEHGB7rj/0Jw/KXsRbGG675OLPP8d4Yw9fY=","timestamp":1588873188673,"extensions":"","signature":"BAMARzBFAiANLSxiimhACoGuUttVPn0V/pP8O7jtYpZowZSmivty1AIhAPeZXrGMOJBrlGjmH41qab7qAnwFBdhVg8Dx7W6Qfv0E"}
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT: begins
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT:  ct_timestamp: 00000171.f038d541
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: timeStampHexStringToByteArray: len =17
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT:  issuer_key_hash.length = 32
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT:  data_len = 715
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService:verifySCT: failed to verify SCT signature
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: verifySCT ends
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService.createSCTextension: ct_id: b0cc83e5a5f97d6baf7c09cc284904872ac7e88b132c6350b7c6fd26e16c6c77
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService.createSCTextension: ct_timestamp: 00000171.f038d2f5
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: timeStampHexStringToByteArray: len =17
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService.createSCTextension: ct_signature: 04030047304502205347f68e7228670cd2955d82cdb5d5897a36130c9953627e7fa0dd011069d1c5022100f317b8602a74a588c095ba1adba9d6017988142ce2e230bffcb249328d9b0df6
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService.createSCTextension: sct_len = 118
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService.createSCTextension: ct_sct_ext id = 1.3.6.1.4.1.11129.2.4.2
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService.createSCTextension: CT extension constructed
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService: issueX509Cert - CT: about to check sct ext
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService: issueX509Cert - CT: SCT ext found added successfully
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CAService: issueX509Cert - CT:About to ca.sign cert.
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: sign cert get algorithm
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: sign cert encoding cert
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: sign cert encoding algorithm
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: CA cert signing: signing cert
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: Getting algorithm context for SHA256withRSA RSASignatureWithSHA256Digest
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: Signing Certificate
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINEST: Property ca.signing.testSignatureFailure not found
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINEST: Getting ca.signing.testSignatureFailure=false
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: storeX509Cert 21
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: In storeX509Cert
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: LdapBoundConnFactory: getting a connection
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: LdapBoundConnFactory: master connection is connected: true
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: LdapBoundConnFactory: connection already connected: true
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: LdapBoundConnFactory: number of connections: 2
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINEST: Getting dbs.newSchemaEntryAdded=true
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINEST: Getting dbs.newSchemaEntryAdded=true
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: DBSSession: add(cn=21,ou=certificateRepository, ou=ca, dc=ca,dc=pki,dc=example,dc=com)
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: LdapBoundConnFactory: number of connections: 3
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] INFO: CA stored signed certificate serial number 0x15
2020-05-07 13:39:49 [https-jsse-nio-8443-exec-15] FINE: done storeX509Cert
````

ladycfu commented on base/ca/src/org/dogtagpki/ct/sct/SCTProcessor.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public List<LogServer> getLogServerConfig() throws EPropertyNotFound, EBaseException, MalformedURLException {
+
```
perhaps for clarity, add a comment with an example set of CS.cfg's CT log config.

SilleBille replied:
Done

ladycfu commented on base/ca/src/org/dogtagpki/ct/LogServer.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void setVersion(int version) {
+        this.version = version;
```
since we do not support version 2 yet, could we disallow anything other than 1 here?

SilleBille replied:
Done

ladycfu commented on base/ca/src/com/netscape/ca/CAService.java:
```diff
@@ -874,76 +878,97 @@ private X509CertImpl issueX509Cert(
+            sct.init();
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+        }
```
I suppose since we are still in "prototype" stage, it's okay if we hold off taking further actions on exception till later.

SilleBille replied:
Done

ladycfu commented on base/ca/src/com/netscape/ca/CAService.java:
```diff
@@ -874,76 +878,97 @@ private X509CertImpl issueX509Cert(
-                } catch (Exception e) {
-                    logger.debug(method + " check sct failed:" + e.toString());
+                        // TODO: Refactor to form right rest API
+                        String ct_uri = ls.getUrl() + "ct/v1/add-pre-chain";
```
I suggest that you add an example CT log config entry in the CS.cfg complete with comments, but with ca.certTransparency.enable set to off by default.  Take a look at ca.scep as example.

SilleBille replied:
Done

SilleBille commented:
Hi @ladycfu Thanks for the review. I have addressed your comments and marked your comments as resolved. Let me know if you have any other comments :smile:

ladycfu commented on base/ca/src/org/dogtagpki/ct/sct/SCTProcessor.java:
```diff
@@ -0,0 +1,103 @@
+     * ca.certTransparency.log.2.pubKey=MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEKATl2B3SAbxyzGOfNRB+AytNTGvdF/FFY6HzWb+/HPE4lJ37vx2nEm99KYUy9SoNzF5VyTwCQG5nL/c5Q77yQQ==
+     * ca.certTransparency.log.2.url=http://ct.googleapis.com:80/logs/crucible/
+     * ca.certTransparency.log.2.version=1
+     * ca.certTransparency.log.num=3
```
Isn't the num supposed to be 1 as only one is enabled?  Question: if you put two, would it still be fine as the 2nd one is not enabled?

SilleBille replied:
Answering here for future reference.

In this particular case,  it can be either 1 or 2. But, in general, `ca.certTransparency.log.num` represents "The total number of CT log entries defined, both enabled and disabled"

SilleBille commented:
Thanks for the review! Merging based on the ACK and green light from CI!

ladycfu commented:
@SilleBille sorry for this late comment, but I just pulled up the latest code and noticed that the pubKey is still using the GoogleTestTube_Pub that I hard-coded in the prototype instead of the CS.cfg parameter
ca.certTransparency.log.1.pubKey

SilleBille commented:
@ladycfu hmmm. Since the `verifySCT()` method wasn't working I missed updating it. Following is a quick diff of the changes required. Feel free to apply the below changes in your HEAD and push it along with your other changes OR I can push a commit tomorrow (should I need to submit a PR for you to review?)

````
diff --git a/base/ca/src/com/netscape/ca/CAService.java b/base/ca/src/com/netscape/ca/CAService.java
index 4b9d79c4c..0189827d6 100644
--- a/base/ca/src/com/netscape/ca/CAService.java
+++ b/base/ca/src/com/netscape/ca/CAService.java
@@ -131,8 +131,6 @@ public class CAService implements ICAService, IService {
     public static final String CHALLENGE_PHRASE = "challengePhrase";
     public static final String SERIALNO_ARRAY = "serialNoArray";
 
-    public static final String GoogleTestTube_Pub = "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEw8i8S7qiGEs9NXv0ZJFh6uuOmR2Q7dPprzk9XNNGkUXjzqx2SDvRfiwKYwBljfWujozHESVPQyydGaHhkaSz/g==";
-
     // CCA->CLA connector
     protected static IConnector mCLAConnector = null;
 
@@ -925,7 +923,7 @@ public class CAService implements ICAService, IService {
                             logger.debug("Response from log server " + respS);
 
                             // verify the sct: TODO - not working, need to fix
-                            verifySCT(CTResponse.fromJSON(respS), cert.getTBSCertificate());
+                            verifySCT(CTResponse.fromJSON(respS), cert.getTBSCertificate(), ls.getPublicKey());
 
                             ctResponses.add(respS);
                         }
@@ -1158,7 +1156,7 @@ public class CAService implements ICAService, IService {
            TBSCertificate tbs_certificate;
          } PreCert;
     */
-    void verifySCT(CTResponse response, byte[] cert)
+    void verifySCT(CTResponse response, byte[] cert, String logPublicKey)
             throws Exception {
 
         String method = "CAService:verifySCT: ";
@@ -1177,7 +1175,7 @@ public class CAService implements ICAService, IService {
         byte[] version = new byte[] {0}; // v1(0)
         byte[] signature_type = new byte[] {0}; // certificate_timestamp(0)
         byte[] entry_type = new byte[] {0, 1}; // LogEntryType: precert_entry(1)
-        byte google_pub[] = CryptoUtil.base64Decode(GoogleTestTube_Pub);
+        byte google_pub[] = CryptoUtil.base64Decode(logPublicKey);
         PublicKey google_pubKey = KeyFactory.getInstance("RSA").generatePublic(
                 new X509EncodedKeySpec(google_pub));
         /*
````
